src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -735,9 +738,30 @@ void WindowEmperor::_createMessageWindow(const wchar_t* className)
     StringCchCopy(_notificationIcon.szTip, ARRAYSIZE(_notificationIcon.szTip), appNameLoc.c_str());
 }
 
+// Counterpart to _postQuitMessageIfNeeded:
+// If it returns true, don't close that last window, if any.
+// This ensures we persist the last window.
+bool WindowEmperor::_shouldSkipClosingWindows() const
+{
+    const auto globalSettings = _app.Logic().Settings().GlobalSettings();
+    const auto windowLimit = globalSettings.ShouldUsePersistedLayout() ? 1 : 0;
+    return _windows.size() <= windowLimit;
+}
+
+// Posts a WM_QUIT as soon as we have no reason to exist anymore.
+// That basically means no windows [^1] and no message boxes.
+//
+// [^1] Unless:
+// * We've been asked to persist the last remaining window
+//   in which case we exit with 1 remaining window.
+// * We're allowed to be headless
+//   in which case we never exit.
 void WindowEmperor::_postQuitMessageIfNeeded() const
 {
-    if (_messageBoxCount <= 0 && _windows.empty() && !_app.Logic().Settings().GlobalSettings().AllowHeadless())
+    const auto globalSettings = _app.Logic().Settings().GlobalSettings();
+    const auto windowLimit = globalSettings.ShouldUsePersistedLayout() ? 1 : 0;

```

Comment:
is it important that we keep  the logic here separate from `_shouldSkipClosingWindows`?

Reply:
You mean I should just call `_shouldSkipClosingWindows` in this function?

Reply:
Hmm I'm a little hesitant to do that, because I'm currently very unhappy with this approach and I don't want to codify this weird dependency. As an example, it doesn't consider the existence of the quake window right now.

Reply:
My gut feeling was correct. This PR is completely broken. 😅